File: typescript/src/TopInterview150/Array-String/RemoveElement/RemoveElement.test.ts
import {removeElement} from './RemoveElement';

describe('Remove Element', () => {
  it('example 1', () => {
    evaluate([3, 2, 2, 3], 3, 2, [2, 2]);
  });
  it('example 2', () => {
    evaluate([0, 1, 2, 2, 3, 0, 4, 2], 2, 5, [0, 1, 4, 0, 3]);
  });
  it('empty array', () => {
    evaluate([], 0, 0, []);
  });
  it('array of size 1: no removals', () => {
    evaluate([1], 0, 1, [1]);
  });
  it('array of size 1: with removal', () => {
    evaluate([1], 1, 0, []);
  });
  it('array of size 2: remove first', () => {
    evaluate([0, 1], 0, 1, [1]);
  });
  it('array of size 2: remove last', () => {
    evaluate([0, 1], 1, 1, [0]);
  });
  it('array of size 3: remove middle', () => {
    evaluate([0, 1, 2], 1, 2, [0, 2]);
  });
});

function evaluate(nums: number[], val: number, expectedOutput: number, expectedNums: number[]): void {
  const numsCopy = [...nums];
  const output = removeElement(numsCopy, val);
  expect(output).toBe(expectedOutput);
  expect(numsCopy.slice(0, output)).toStrictEqual(expectedNums);
}

File: scala/src/test/scala/jbaldassari/leetcode/topinterview150/arraystring/removeelement/RemoveElementTest.scala
package jbaldassari.leetcode.topinterview150.arraystring.removeelement

class RemoveElementTest extends munit.FunSuite {
  test("example 1") {
    evaluate(Array(3, 2, 2, 3), 3, 2, Array(2, 2))
  }
  test("example 2") {
    evaluate(Array(0, 1, 2, 2, 3, 0, 4, 2), 2, 5, Array(0, 1, 4, 0, 3))
  }
  test("empty array") {
    evaluate(Array(), 0, 0, Array())
  }
  test("array of size 1: no removals") {
    evaluate(Array(1), 0, 1, Array(1))
  }
  test("array of size 1: with removal") {
    evaluate(Array(1), 1, 0, Array())
  }
  test("array of size 2: remove first") {
    evaluate(Array(0, 1), 0, 1, Array(1))
  }
  test("array of size 2: remove last") {
    evaluate(Array(0, 1), 1, 1, Array(0))
  }
  test("array of size 3: remove middle") {
    evaluate(Array(0, 1, 2), 1, 2, Array(0, 2))
  }
  def evaluate(nums: Array[Int], valueToRemove: Int, expectedOutput: Int, expectedNums: Array[Int]): Unit = {
    val numsCopy = nums.clone()
    val result = Solution.removeElement(numsCopy, valueToRemove)
    assertEquals(result, expectedOutput)
    assertEquals(numsCopy.toList.take(expectedNums.size), expectedNums.toList)
  }
}
